[CI/BUILD] enable intel queue for longer CPU tests

zhouyuan opened the pull request:
FILL IN THE PR DESCRIPTION HERE

This PR enables intel queue for longer CPU tests. The main changes are
- fix the bug on pos_encoding kernel 
- enable part of the tests under `models`
- skip for tests for cuda only on CPU

related: #3654 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

zhouyuan committed:
[CI] enable intel queue for longer CPU tests

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
enable more tests

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
ignore models

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
adding big model test

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
disable big model

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

simon-mo commented on .buildkite/run-cpu-test.sh:
```diff
@@ -11,4 +11,25 @@ trap remove_docker_container EXIT
+
+# Run basic model test
+docker exec cpu-test bash -c "cd tests;
+  pip install pytest Pillow
+  rm -f __init__.py
+  sed -i '/\"stabilityai/d' models/test_models.py
+  sed -i '/torch.cuda.empty_cache/d' conftest.py
+  sed -i 's/cuda/cpu/g' conftest.py
```
please make the corresponding changes properly, don't do this patch.

zhouyuan committed:
add device check in conftest

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan commented on tests/models/test_models.py:
```diff
@@ -14,7 +14,6 @@
     "EleutherAI/pythia-70m",
     "bigscience/bloom-560m",  # Testing alibi slopes.
     "microsoft/phi-2",
-    "stabilityai/stablelm-3b-4e1t",
```
the model "stabilityai/stablelm" seems requireing some special handling on pos_encoding, which is not implemented yet . Will fix this in later commits

zhouyuan commented on .buildkite/run-cpu-test.sh:
```diff
@@ -11,4 +11,23 @@ trap remove_docker_container EXIT
+# Run basic model test
+docker exec cpu-test bash -c "cd tests;
+  pip install pytest Pillow
+  rm -f __init__.py
```
need this line otherwise this will trigger some missing module issue: `Import vllm._C error: No module named 'vllm._C'`

zhouyuan replied:
https://stackoverflow.com/questions/41748464/pytest-cannot-import-module-while-python-can

bigPYJ1151 committed:
Fix corner case in pos_encoding.

zhouyuan committed:
enable stabilityai model

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan commented on .buildkite/run-cpu-test.sh:
```diff
@@ -11,4 +11,22 @@ trap remove_docker_container EXIT
+# offline inference
+docker exec cpu-test bash -c "python3 examples/offline_inference.py"
+
+# async engine test
```
pending on #3993

jikunshang commented on tests/conftest.py:
```diff
@@ -51,7 +52,8 @@ def cleanup():
         torch.distributed.destroy_process_group()
     gc.collect()
-    torch.cuda.empty_cache()
+    if not is_cpu():
```
better use  `if torch.cuda.is_available()`

jikunshang commented on tests/conftest.py:
```diff
@@ -136,6 +138,12 @@ def example_long_prompts() -> List[str]:
+    def wrap_device(self, input: any):
+        if is_cpu():
+            return input.cpu()
+        else:
```
same as above

zhouyuan committed:
using torch.cuda.is_available() to check the device type

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
fix format

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
ignore failed tests firstly

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
ignore embedding test

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
fix test dir

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
fix tests with cuda device only

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
add local develop

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
fix format

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
soft link tests folder

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
enable big model test

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
addning hf token

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
enable test mistral

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan commented:
@simon-mo hi Simon, I made more progress and covered more model tests for CPU backend, please kindly help to check and advise.

thanks
-yuan

simon-mo commented on .buildkite/run-cpu-test.sh:
```diff
@@ -10,5 +10,22 @@ remove_docker_container() { docker rm -f cpu-test || true; }
+# Run big model test
+docker exec cpu-test bash -c "
+  #TODO: remove this after CPU float16 support ready
+  sed -i 's/half/float/g' tests/models/test_big_models.py
```
I think this should be inside the code. You can change `test_big_models.py` to conditionally use `float` on CPU and `half` for GPU tests, and add a comment.

zhouyuan replied:
thanks, good point, fixed in https://github.com/vllm-project/vllm/pull/4113/commits/f3e300ea37e4b08912b453c5af665d1d8d1cd481

simon-mo reviewed (commented):
This looks good to me overall pending `test_big_models.py` to be fixed. If the test suite grow, we can take the same approach as AMD did for `mirror_hardwares` to run all tests in Intel CPUs.

zhouyuan committed:
use float dtype for big model tests with CPU backend

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan commented on Dockerfile.cpu:
```diff
@@ -19,4 +19,6 @@ RUN VLLM_TARGET_DEVICE=cpu python3 setup.py install
 WORKDIR /workspace/
 
+RUN ln -s /workspace/vllm/tests  && ln -s /workspace/vllm/examples && ln -s /workspace/vllm/benchmarks
+
```
https://github.com/vllm-project/vllm/pull/4971

bigPYJ1151 commented on .buildkite/run-cpu-test.sh:
```diff
@@ -10,5 +10,15 @@ remove_docker_container() { docker rm -f cpu-test || true; }
-# Run the image and launch offline inference
-docker run --network host --env VLLM_CPU_KVCACHE_SPACE=1 --name cpu-test cpu-test python3 vllm/examples/offline_inference.py
+# Run the image
+docker run -itd -v ~/.cache/huggingface:/root/.cache/huggingface --network host -e HF_TOKEN --env VLLM_CPU_KVCACHE_SPACE=4 --name cpu-test cpu-test
```
Maybe binding the container to a numa node will make the tests faster.
```suggestion
docker run -itd -v ~/.cache/huggingface:/root/.cache/huggingface --network host --cpuset-cpus=48-79 --cpuset-mems=1 -e HF_TOKEN --env VLLM_CPU_KVCACHE_SPACE=4 --name cpu-test cpu-test
```

zhouyuan commented:
> This looks good to me overall pending `test_big_models.py` to be fixed. If the test suite grow, we can take the same approach as AMD did for `mirror_hardwares` to run all tests in Intel CPUs.

@simon-mo just fixed the `test_big_models.py` issue and thanks for the inputs! Will looking into that direction

thanks
-yuan

zhouyuan commented:
@simon-mo 
Hi, the CI failed after several retries, it may be related with https://github.com/vllm-project/vllm/pull/5091 so I did a rebase with the main branch to try again

thanks, -yuan

zhouyuan committed:
fix failed CI on CUDA spawn issue

torch.cuda.is_available() is not working well with multiprocessing case
so we switch to use is_cpu() to check the device

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan commented:
@simon-mo Hi, the CI for CUDA backend is fixed now. The issue is due to` cuda.is_available()` not working well when using `multiprocessing` to launch other process. 
The AMD CI is failing today due to `Unable to open /dev/kfd read-write: Invalid argument`, looks like an environment issue

thanks
-yuan

zhouyuan committed:
fix rebase issue

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
fix rebase

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
Trigger CI

zhouyuan commented:
@simon-mo thanks a lot for review and merge, will move on to cover more tests on CPU backend.

thanks, -yuan
